File: Borland Integration/C++5/Integrator.cpp
// ---------------------------
// (c) Reliable Software, 2002
// ---------------------------
#include "Integrator.h"
#include "SccProxy.h"
#include <StringOp.h>
#include <Ex/WinEx.h>
#include <File/File.h>
#include <Sys/WinString.h>
#include <LightString.h>
#include <iostream>
#include <assert.h>

char Integrator::CodeCoopExpert [] = "Reliable Software.Code Co-op Expert";

// IUnknown
HRESULT STDMETHODCALLTYPE Integrator::QueryInterface (
            REFIID riid,
            void __RPC_FAR *__RPC_FAR *ppvObject)
{
    *ppvObject = NULL;

    if (IsEqualIID (riid, __uuidof (IUnknown)))
        *ppvObject = static_cast<IUnknown *> (this);
    else if (IsEqualIID (riid, __uuidof (IOTANotifier)))
        *ppvObject = static_cast<IOTANotifier *> (this);
    else if (IsEqualIID (riid, __uuidof (IOTAWizard)))
        *ppvObject = static_cast<IOTAWizard *> (this);

    if (*ppvObject != NULL)
    {
        reinterpret_cast<IUnknown *>(*ppvObject)->AddRef ();
        return S_OK;
    }
    else
        return E_NOINTERFACE;
}

ULONG STDMETHODCALLTYPE Integrator::AddRef ()
{
    return ::InterlockedIncrement (&_refCount);
}

ULONG STDMETHODCALLTYPE Integrator::Release ()
{
  if (::InterlockedDecrement (&_refCount) == 0)
  {
    delete this;
    return 0;
  }
  return _refCount;
}

// IOTAWizard
void __fastcall Integrator::Execute ()
{
    assert (!"never called");
}

AnsiString __fastcall Integrator::GetIDString ()
{
    return CodeCoopExpert;
}

AnsiString __fastcall Integrator::GetName ()
{
    return CodeCoopExpert;
}

TWizardState __fastcall Integrator::GetState ()
{
    return TWizardState() << wsEnabled;
}

Integrator::Integrator ()
: _refCount (0)
{
    // Create Coop menu
    _coopExe.reset (NewItem ("Launch &Code Co-op", 0, false, true, OnLaunchCoop, 0, "CoopExe"));
    _checkOut.reset (NewItem ("Check-&Out", 0, false, true, OnCheckOut, 0, "CheckOut"));
    _separator.reset (NewLine ());
    _about.reset (NewItem ("&About...", 0, false, true, OnAbout, 0, "About"));

    TMenuItem * items [4];
    items [0] = _coopExe.get ();
    items [1] = _checkOut.get ();
    items [2] = _separator.get ();
    items [3] = _about.get ();

    _coopMenu.reset (NewSubMenu ("C&ode Co-op", 0, "CodeCoop", items, 3, true));
    _coopMenu->OnClick = OnClickCoopMenu;

    // Install the menu into main IDE menu
    _di_INTAServices ntaServices;
    // Revisit: Handle exceptions
    BorlandIDEServices->QueryInterface (
                                __uuidof (INTAServices),
                                (void **)&ntaServices);

    TMainMenu * mainMenu = ntaServices->GetMainMenu ();
    mainMenu->Items->Insert (mainMenu->Items->Count - 2, _coopMenu.get ());

}

Integrator::~Integrator ()
{
    _di_INTAServices ntaServices;
    if (SUCCEEDED (BorlandIDEServices->QueryInterface (IID_INTAServices, (void**)&ntaServices)))
    {
        TMainMenu * mainMenu = ntaServices->GetMainMenu ();
        mainMenu->Items->Remove (_coopMenu.get ());
    }
}

void __fastcall Integrator::OnClickCoopMenu (TObject *Sender)
{
    try
    {
        _currentFile.clear ();
        // get the name of the current file
        _di_IOTAModuleServices ms;
        HRESULT hr = BorlandIDEServices->QueryInterface (
                            __uuidof (IOTAModuleServices),
                           (void **)&ms);
        if (FAILED (hr))
        {
            // Revisit: use exception that understands HRESULT values
            throw Win::Exception ("Cannot query interface", "IOTAModuleServices");
        }
        if (ms != NULL)
        {
          _di_IOTAModule m = ms->CurrentModule ();
          if (m != NULL)
              _currentFile = m->GetFileName ().c_str ();
        }
        if (_currentFile.empty ())
        {
            _checkOut->Caption = "Check-Out";
            _checkOut->Enabled = false;
        }
        else
        {
            _checkOut->Enabled = true;
            _checkOut->Caption = "Check-Out ";
            _checkOut->Caption += _currentFile.c_str ();
        }
    }
    catch (Win::Exception e)
    {
        DisplayEx ("On Menu Click", e);
    }
    catch (...)
    {
    	Win::ClearError ();
        ::MessageBox (0, "On Menu Click: Unknown problem", "Code Co-op", MB_OK);
    }
}

void __fastcall Integrator::OnLaunchCoop (TObject * Sender)
{
    try
    {
    	SccProxy sccProxy;
        if (!sccProxy.ExecuteCoop ())
	    throw Win::Exception ("Cannot execute Code Co-op.");
    }
    catch (Win::Exception e)
    {
        DisplayEx ("Launch Code Co-op", e);
    }
    catch (...)
    {
    	Win::ClearError ();
        ::MessageBox (0, "Launch Co-op: Unknown problem", "Code Co-op", MB_OK);
    }
}

void __fastcall Integrator::OnCheckOut (TObject * Sender)
{
    // in BCB together with cpp file check out also
    // 1) corresponding header file
    // 2) corresponding dfm file
    //
    // in Delphi together with pas file check out also
    // 1) corresponding dfm file

    Assert (!_currentFile.empty ());
    try
    {
        // check-out
        char const * curFile = _currentFile.c_str ();

        SccProxy sccProxy;
        if (!sccProxy.CheckOut (1, &curFile, 0))
            throw Win::Exception ("Cannot checkout the current file.", curFile);

        MakeReadWriteInIDE ();

		PathSplitter splitter (curFile);
        bool isCpp = IsNocaseEqual (".cpp", splitter.GetExtension ());
		if (isCpp || IsNocaseEqual (".pas", splitter.GetExtension ()))
		{
            std::string currentHdr;
			std::string currentDfm = splitter.GetDrive ();
            currentDfm += splitter.GetDir ();
            currentDfm += splitter.GetFileName ();
            currentHdr = currentDfm;
            currentDfm += ".dfm";
            currentHdr += ".h";

			char const * dfm = currentDfm.c_str ();
			if (File::Exists (dfm))
            {
        		if (!sccProxy.CheckOut (1, &dfm, 0))
            		throw Win::Exception ("Cannot checkout the current file.", dfm);
            }
            if (!isCpp)
                return;

   	 		char const * hdr = currentHdr.c_str ();
    		if (!File::Exists (hdr))
                return;

           	if (!sccProxy.CheckOut (1, &hdr, 0))
           		throw Win::Exception ("Cannot checkout the current file.", hdr);
        }
    }
    catch (Win::Exception e)
    {
        DisplayEx ("Checkout", e);
    }
    catch (...)
    {
    	Win::ClearError ();
        ::MessageBox (0, "Checkout: Unknown problem", "Code Co-op", MB_OK);
    }
}

void Integrator::MakeReadWriteInIDE ()
{
    // make the current editor's file read-write to IDE
    /*
        first way

        _di_IOTAActionServices as;
        hr = BorlandIDEServices->QueryInterface (
                            __uuidof (IOTAActionServices),
                           (void **)&as);
        if (FAILED (hr))
        {
            // Revisit: use exception that understands HRESULT values
            throw Win::Exception ("Cannot query interface", "IOTAActionServices");
        }

        as->ReloadFile (curFile);
    */

    //   second way, faster but hackish:
    if (!File::IsReadOnly (_currentFile.c_str ())) // file is r-w on disk
    {
        _di_IOTAEditorServices es;
        HRESULT hr = BorlandIDEServices->QueryInterface (
                            __uuidof (IOTAEditorServices),
                           (void **)&es);
        if (FAILED (hr))
        {
            // Revisit: use exception that understands HRESULT values
            throw Win::Exception ("Cannot query interface", "IOTAEditorServices");
        }

        es->TopBuffer->SetIsReadOnly (False);
    }
}

void __fastcall Integrator::OnAbout (TObject * Sender)
{
    static char const aboutTxt [] =
                "This is Code Co-op\nServer-less Version Control System\n"
                "\nby Reliable Software.\n\nVisit our web site at:\nwww.ReliSoft.com";
    ::MessageBox (0, aboutTxt, "Code Co-op Version Control System", MB_OK);
}

void Integrator::DisplayEx (char const * cmd, Win::Exception const & e) const
{
    Msg msg;
    msg << cmd << ": " << e.GetMessage () << std::endl;
    SysMsg sysMsg (e.GetError ());
    if (sysMsg)
        msg << "System tells us: " << sysMsg.Text ();

    std::string objectName (e.GetObjectName ());
    if (!objectName.empty ())
      	msg << "    " << objectName << std::endl;

    ::MessageBox (0, msg.c_str (), "Code Co-op", MB_OK);
}

File: Borland Integration/Delphi5/integrator.pas
// ---------------------------
// (c) Reliable Software, 2001
// ---------------------------

unit integrator;

interface

uses
  Windows, ToolsAPI, Registry, Menus, Forms, SysUtils;

resourcestring CodeCoopExpert = 'Reliable Software.Code Co-op Expert';
resourcestring SccDll = '\sccdll.dll';
resourcestring AboutTxt = 'This is Code Co-op'#13#10'Version Control System'#13#10'by Reliable Software.'#13#10#10'Visit our web site at:'#13#10'www.ReliSoft.com';

type
  TIntegrator = class (TNotifierObject, IOTANotifier, IOTAWizard)
  public
    constructor Create;
    destructor Destroy; override;

    // IOTAWizard
    function GetIDString: string;
    function GetName: string;
    function GetState: TWizardState;
    procedure Execute ();
    // IOTANotifier
    procedure AfterSave;
    procedure BeforeSave;
    procedure Destroyed;
    procedure Modified;

  protected
    procedure OnClickCoopMenu (Sender : TObject);
    procedure OnClickLaunchCoop (Sender : TObject);
    procedure OnClickCheckout (Sender : TObject);
    procedure OnClickAbout (Sender : TObject);
  private
    _coopMenu : TMenuItem;
    _coopExe, _checkOut, _sep, _about : TMenuItem;

    _programPath : string;
    _currentFile : AnsiString;
    _notifierIndex : Integer;
end;

implementation

{ TIntegrator }

constructor TIntegrator.Create;
var
    mainMenu : TMainMenu;
    coopRegistry : TRegistry;
begin
    inherited Create;

    _programPath := '';
    coopRegistry := TRegistry.Create;
    try
        coopRegistry.RootKey := HKEY_LOCAL_MACHINE;
        if coopRegistry.OpenKey ('\Software\Reliable Software\Code Co-op', False) then
        begin
            _programPath := coopRegistry.ReadString ('ProgramPath');
            coopRegistry.CloseKey;
        end
    finally
        coopRegistry.Free;
        inherited;
    end;
    if (_programPath = '') then
    begin
        Application.MessageBox (
            'Cannot read Code Co-op registry. Run setup again.',
            'Code Co-op...', IDOK);
    end;

    // Create drop down menu
    _coopMenu := TMenuItem.Create (nil);
    _coopMenu.Caption := 'C&ode Coop';
    _coopMenu.Name    := 'CodeCoop';
    _coopMenu.OnClick := OnClickCoopMenu;

    // Insert drop down menu.
    mainMenu := (BorlandIDEServices as INTAServices).MainMenu;
    Assert(Assigned(mainMenu), 'MainMenu component not found');
    mainMenu.Items.Insert (mainMenu.Items.Count - 2, _coopMenu);

    _coopExe := TMenuItem.Create (_coopMenu);
    _coopExe.Caption := 'Launch Code Co-Op';
    _coopExe.Name := 'CodeCoop';
    _coopExe.OnClick := OnClickLaunchCoop;
    _coopExe.Enabled := _programPath <> '';
    _coopMenu.Add (_coopExe);

    _checkOut := TMenuItem.Create (_coopMenu);
    _checkOut.Caption := 'Check out';
    _checkOut.Name := 'Checkout';
    _checkOut.OnClick := OnClickCheckout;
    _checkOut.Enabled := _programPath <> '';
    _coopMenu.Add (_checkOut);

    _sep := TMenuItem.Create (_coopMenu);
    _sep.Caption := '-';
    _sep.Name := 'Separator';
    _coopMenu.Add (_sep);

    _about := TMenuItem.Create (_coopMenu);
    _about.Caption := 'About...';
    _about.Name := 'About';
    _about.OnClick := OnClickAbout;
    _about.Enabled := True;
    _coopMenu.Add (_about);

{
   Application.MessageBox ('Before add notifier', 'Add Notifier');
  _notifierIndex := (BorlandIDEServices as IOTAServices).AddNotifier (Self);
   Application.MessageBox ('After add notifier', 'Add Notifier');
  if _notifierIndex < 0 then
    Application.MessageBox ('Failed!', 'Add Notifier');
}

end;

destructor TIntegrator.Destroy;
begin
  _coopMenu.Free;
//  (BorlandIDEServices as IOTAServices).RemoveNotifier (_notifierIndex);

  inherited Destroy;
end;

// IOTAWizard
function TIntegrator.GetIDString: string;
begin
  Result := CodeCoopExpert;
end;

function TIntegrator.GetName: string;
begin
  Result := CodeCoopExpert;
end;

function TIntegrator.GetState: TWizardState;
begin
  Result := [wsEnabled];
end;

procedure TIntegrator.Execute ();
begin
    Assert (false); // never called !
end;

// IOTANotifier
procedure TIntegrator.AfterSave;
begin
    Application.MessageBox ('Hey, watch out!', 'After save', IDOK);
end;

procedure TIntegrator.BeforeSave;
begin
    Application.MessageBox ('Hey, watch out!', 'Before save', IDOK);
end;

procedure TIntegrator.Destroyed;
begin
    Application.MessageBox ('Hey, watch out!', 'Destroyed', IDOK);
end;

procedure TIntegrator.Modified;
begin
    Application.MessageBox ('Hey, watch out!', 'Modified', IDOK);
end;

// private
procedure TIntegrator.OnClickCoopMenu (Sender : TObject);
var
    moduleServices : IOTAModuleServices;
    currentModule : IOTAModule;

begin
    if (_programPath = '') then
        Exit;

    _currentFile := '';

    moduleServices := BorlandIDEServices as IOTAModuleServices;
    currentModule := moduleServices.CurrentModule ();

    if Assigned (currentModule) then
        _currentFile := currentModule.FileName;

    if (_currentFile <> '') then
    begin
        _checkOut.Enabled := True;
        _checkOut.Caption := 'Check-Out ' + _currentFile;
    end
    else
    begin
        _checkOut.Caption := 'Check-Out';
        _checkOut.Enabled := False;
    end;
end;

procedure TIntegrator.OnClickLaunchCoop (Sender : TObject);
type
    TSccRunScc = function (context : PAnsiChar;
                           hWnd : PInteger;
                           nFiles : Integer;
                           lpFileNames : PAnsiString) : Boolean; cdecl;

var
  h: Integer;
  sccRun : TSccRunScc;

begin
     Assert (_programPath <> '');

     h := LoadLibrary (PChar (_programPath + SccDll));
     if h <> 0 then
     begin
         @sccRun := GetProcAddress(h, 'SccRunScc');
         if @sccRun <> nil then
         begin
             sccRun (NIL,NIL,0,NIL);
         end;
         FreeLibrary(h);
     end;
end;

procedure TIntegrator.OnClickCheckout (Sender : TObject);
type
    TSccCheckout = function (context : PAnsiChar;
                                hWnd : Integer;
				nFiles : Integer;
				lpFileNames : PAnsiString;
                                lpComment : PAnsiString;
                                dwFlags : Integer;
                                pvOptions : PChar) : Boolean; cdecl;

var
    h : Integer;
    sccCheckout : TSccCheckout;
    dfmCurrentFile : AnsiString;
    attr : DWORD;
    fnameLen : Integer;
begin
     Assert (_programPath <> '');
     Assert (_currentFile <> '');

     h := LoadLibrary (PChar (_programPath + SccDll));
     if h <> 0 then
     begin
         @sccCheckout := GetProcAddress(h, 'SccCheckout');
         if @sccCheckout <> nil then
         begin
             sccCheckout (NIL, 0, 1, @_currentFile, NIL, 0, NIL);
             // for '*.pas' or '*.cpp' files also checkout corresponding '*.dfm' file
             fnameLen := Length (_currentFile) - 4;
             if ((AnsiPos ('.pas', _currentFile) - 1 = fnameLen) or
                 (AnsiPos ('.cpp', _currentFile) - 1 = fnameLen)) then
             begin
                 dfmCurrentFile := Copy (_currentFile, 0, Length (_currentFile) - 3);
                 dfmCurrentFile := Concat (dfmCurrentFile, 'dfm');
                 attr := GetFileAttributes (PChar (dfmCurrentFile));
                 if (attr <> $ffffffff) then
                     sccCheckout (NIL, 0, 1, @dfmCurrentFile, NIL, 0, NIL);
             end;
         end;
         FreeLibrary(h);
     end;

//   (BorlandIDEServices as IOTAActionServices).ReloadFile (currentFile);
//    or faster but hackish:

        attr := GetFileAttributes (PChar (_currentFile));
        if (attr <> $ffffffff) then
          if ((attr and FILE_ATTRIBUTE_READONLY) = 0) then
            (BorlandIDEServices as IOTAEditorServices).TopBuffer.SetIsReadOnly (False);
end;

procedure TIntegrator.OnClickAbout (Sender : TObject);
begin
    Application.MessageBox (PChar (AboutTxt), 'About...', IDOK);
end;

end.


